app/controllers/sessions_controller.rb

```diff
@@ -45,6 +45,8 @@ def find_user(session)
     password = session[:password].is_a?(String) && session.fetch(:password)
 
     User.authenticate(who, password) if who && password
+  rescue BCrypt::Errors::InvalidHash
+    nil
```

Comment:
We call sign_in/authenticate from profile update, delete and API key show as well.
I would suggest adding this in application controller as `rescue_from BCrypt::Errors::InvalidHash, with: render_invalid_password` which should support both html and plain format. Adding `login.failure.invalid_password` would be nice as well ([example](https://github.com/rubygems/rubygems.org/blob/master/app/controllers/sessions_controller.rb#L38)).

Reply:
> We call sign_in/authenticate from profile update, delete and API key show as well.

We don't need to worry profile update/delete as the user needs to already be signed in for these actions, but API keys does need the change.

> I would suggest adding this in application controller as 

We're mimicking behaviour as if the user entered an incorrect password so that clearance can handle the response correctly and not have to worry about it ourselves.

> Adding login.failure.invalid_password

Same here. We're mimicking behaviour as if the user entered an incorrect password, so that behaviour is already implemented.

Reply:
> We're mimicking behaviour as if the user entered an incorrect password

I personally would prefer to keep this separate from incorrect password and show users why exactly their correct password is no longer valid and what do they need to do to fix it. A lot of users who login in 2018 (and a couple in 2019-20) seem to have sha1 passwords, even tho we had the migration setting in use for over 7 years:
```
User.where("encrypted_password ~ ?", '[a-f0-9]{40}').where("remember_token_expires_at > '2018-01-01'").count
=> 93
```
It would be nice if we could monitor this separately and let users know how they can fix this. Nice to have, not a blocker, we can always deal with any confusion through the help site.